File: commons-work-api/src/main/java/io/github/cooperlyt/mis/work/data/WorkInstance.java
package io.github.cooperlyt.mis.work.data;

import lombok.*;

@EqualsAndHashCode(callSuper = true)
@Data
@NoArgsConstructor
public abstract class WorkInstance<T> extends WorkImpl{

  public WorkInstance(WorkInfo workInfo, T instance) {
    super(workInfo.getDefineId(), workInfo.getType(),
        workInfo.isProcess(), workInfo.getWorkId(),
        workInfo.getWorkName(), workInfo.getStatus(),
        workInfo.getDataSource(), workInfo.getCreatedAt(),
        workInfo.getUpdatedAt(), workInfo.getCompletedAt(), workInfo.getValidateAt());
    this.instance = instance;
  }

  private T instance;

  public abstract String getDescription();
}

File: commons-work-api/src/main/java/io/github/cooperlyt/mis/work/data/WorkInstance.kt
package io.github.cooperlyt.mis.work.data

import io.github.cooperlyt.mis.work.data.WorkInfo

interface WorkInstance<T>: WorkInfo {

  val instance: T
  
  val description: String


  data class SimpleWorkInfo(
    private val workInfo: WorkInfo,
    override val instance: String
  ): WorkInstance<String>, WorkInfo by workInfo {
    override val description: String = instance
  }

}